sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala

```diff
@@ -143,8 +143,12 @@ final class Decimal extends Ordered[Decimal] with Serializable {
     } else if (decimal.scale < 0 && !SQLConf.get.allowNegativeScaleOfDecimalEnabled) {
       this._precision = decimal.precision - decimal.scale
       this._scale = 0
-      // set scale to 0 to correct unscaled value
-      this.decimalVal = decimal.setScale(0)
+      if (this._precision > 309) {
```

Comment:
`309` to support double. or it is should be `DecimalType.MAX_PRECISION`?

Reply:
I think the max precision is DecimalType.MAX_PRECISION (38)